index.rst

```diff
@@ -1,57 +1,236 @@
 
-     Caption text.
+Latencies and overheads must be minimized.
+Since the Alert Production timeline is only 60 seconds from close of shutter on the last exposure of the visit, latencies on the order of seconds and potentially tenths of seconds do matter.
+As much I/O and computation as possible should be done after ``next_visit`` arrives before the exposure images do.
+Ingest of image metadata and starting pipelines should be as efficient as possible.
 
-   Run: ``make html`` and ``open _build/html/index.html`` to preview your work.
-   See the README at https://github.com/lsst-sqre/lsst-technote-bootstrap or
-   this repo's README for more info.
+The framework should execute in parallel, with a minimum of one process per CCD/detector.
+Amplifier-level or Source-level parallelism may be desirable for some functions.
 
-   Feel free to delete this instructional comment.
+It should handle single and multi-exposure visits, with the number of snaps per visit ideally known only at ``next_visit`` time.
 
-:tocdepth: 1
+It must be able to execute different pipelines on different images, at a minimum via configuration but ideally on a per-visit basis at ``next_visit`` time.
 
-.. Please do not modify tocdepth; will be fixed when a new Sphinx theme is shipped.
+Since pipeline execution takes longer than the interval between visits, the framework must be able to have more than one visit in flight at a time.
+As some visits may take longer than others due to crowding, artifacts, or other influences, it must be elastic to allow compute resources to be used as needed.
 
-.. sectnum::
+It must be resilient to failures of communication, aborted visits, loss of exposures, memory exhaustion, and other problems.
+But occasional delays in executing or even outright failures to execute the pipeline for a visit are acceptable, as daytime catch-up processing is planned.
 
-.. TODO: Delete the note below before merging new content to the main branch.
 
-.. note::
+Initial Design
+==============
 
-   **This technote is not yet published.**
+The DDS ``next_visit`` event is translated to an HTTP POST containing the same information.
+One POST message is issued per CCD.
+That message is used to reserve a worker to process the entire visit.
+The HTTP message handler does not return until the visit has been successfully processed or a failure has occurred, been identified, and reported.
+
+Within the message handler, a separate messaging system is configured to receive notifications of exposure image arrival for the visit.
+While waiting for those exposures, pipeline-specific calibration data and association data are preloaded into a newly-created handler-local repo.
+As exposures arrive, they are ingested into the local repo, including copying of the image data.
+When the last exposure has arrived, a pipeline selected by one or more fields in the ``next_visit`` event is executed via the Middleware ``pipetask`` command.
```

Comment:
Sorry, overlooked this one because I originally thought we needed all-detector processing.

If the pipelines are truly compartmentalized into different workers, is there any purpose to waiting for the last exposure? I would think you would need a separate `pipetask` (or equivalent) execution on each worker, in which case there is no benefit to holding ready workers while other exposures are still being downloaded.

Reply:
It's not the last (detector x exposure) image of the _focal plane_, it's the last (detector x exposure) image of the _visit_, as we need to combine the snaps in the pipeline.  I.e. worker A processes detector 0, waiting for and retrieving snap images 0 and 1 for that detector, but not any other detector's images.